### PR TITLE
CellH5 reader update and CellH5Writer

### PR DESCRIPTION
**Pull request by sommerc**
I tried the cellh5 reader in fiji with the newest version 5.1 and basically it worked, however, some changes in the RoiHandler made it unusable. Tried to fix these issues and happy for comments.

Cheers,
Christoph

### CODE REVIEW

**Comment by jburel** [comment]
* Change in ROI handler will affect other part of the code assuming that all items are selected, windows is always on top. 
cc @bramalingam

**Comment by sommerc** [comment]
Yes, in my case, loading thousands of ROIs that way leads to a unresponsive fiji that triggered a display update after reading each ROI... Perhaps, just disabling "showAll" could help. Didn't try..

**Comment by bramalingam** [comment]
Yes, that part of the code had to be outside the loop. Thank you for pointing that out.

File: components/formats-gpl/src/loci/formats/in/CellH5Reader.java [hunk 1]
```diff
@@ -577,7 +580,7 @@ private void parseROIs(int s) {
             cellObjectName + " " + objectLabelId, roiManagerRoiIndex);
 
           String shapeID =
-            MetadataTools.createLSID("Shape", roiManagerRoiIndex, 0);
+            MetadataTools.createLSID(cellObjectName, roiManagerRoiIndex, 0);
```
Comment (melissalinkert) [inline components/formats-gpl/src/loci/formats/in/CellH5Reader.java]: This is unfortunately going to cause the OME-XML to be invalid - the ID must be of the format ```Shape:*:*```.

It might be better to call ```store.setRectangleText(cellObjectName, roiManagerRoiIndex, 0)``` in the reader, and then update ROIHandler to set the ImageJ Roi name to ```shapeObject.getText()``` on line 308 (falling back to ```shapeObject.getID()``` if the text is null).  That way the OME-XML remains valid, but the descriptive name shows up in the ROI manager.

File: components/formats-gpl/src/loci/formats/in/CellH5Reader.java [hunk 2]
```diff
@@ -586,11 +589,11 @@ private void parseROIs(int s) {
...
 
           store.setRectangleTheT(
-            new NonNegativeInteger(roiTime), roiManagerRoiIndex, 0);
+            new NonNegativeInteger(roiTime + 1), roiManagerRoiIndex, 0);
           store.setRectangleTheC(
-            new NonNegativeInteger(roiChannel), roiManagerRoiIndex, 0);
+            new NonNegativeInteger(roiChannel + 1), roiManagerRoiIndex, 0);
           store.setRectangleTheZ(
-            new NonNegativeInteger(roiZSlice), roiManagerRoiIndex, 0);
+            new NonNegativeInteger(roiZSlice + 1), roiManagerRoiIndex, 0);
```
Comment (melissalinkert) [inline components/formats-gpl/src/loci/formats/in/CellH5Reader.java]: 0-based indexes are correct here (see e.g. http://www.openmicroscopy.org/Schemas/Documentation/Generated/OME-2015-01/ROI_xsd.html#Shape_TheZ).  If ImageJ is expecting the index to start from 1, then this should be fixed in ROIHandler instead.

File: components/formats-gpl/src/loci/formats/in/CellH5Reader.java [hunk 3]
```diff
@@ -481,6 +495,20 @@ private void parseStructure() throws FormatException {
...
+      store.setPlateName(seriesPlate.get(s), 0);
+      
+      String well_id =  MetadataTools.createLSID("Well", 0);
+      
+      store.setWellID(well_id, 0, 0);
+      store.setWellExternalIdentifier(seriesWell.get(s), 0, 0);
+      
+      String site_id = MetadataTools.createLSID("WellSample", 0);
+      store.setWellSampleID(site_id, 0, 0, 0);
+      store.setWellSampleIndex(NonNegativeInteger.valueOf(seriesSite.get(s)), 0, 0, 0);
```
Comment (melissalinkert) [inline components/formats-gpl/src/loci/formats/in/CellH5Reader.java]: The Well and WellSample population here is potentially problematic.

The row and column values on Well (```store.setWellRow(...)``` and ```store.setWellColumn(...)```) are required to produce valid OME-XML; these should be set after the Well ID.  Only one Well and WellSample are created, even if multiple wells and sites exist - this means that only the last well and site will be stored in the OME-XML.

While not strictly necessary, it's a good idea to also call  ```store.setWellSampleImageRef(...)```, as this ensures that the Image metadata and the Plate/Well metadata are correctly linked together.

An example of populating a full plate is here:

https://github.com/openmicroscopy/bioformats/blob/develop/components/formats-gpl/src/loci/formats/in/ScanrReader.java#L598 through line 650

Comment (sommerc) [inline components/formats-gpl/src/loci/formats/in/CellH5Reader.java]: * We currently have the well information as a alpha numeric coordinate (e.g. B03). WellRow and WellColumn are stored as integers. I can implement a mapping (e.g. C12 will be 3x12; 1-based indexing?) Would this be okay?

* using store.setWellSampleImageRef() makes totally sense to me

thanks for all your help!

File: components/formats-gpl/src/loci/formats/in/CellH5Reader.java [hunk 4]
```diff
@@ -481,8 +495,44 @@ private void parseStructure() throws FormatException {
...
+          store.setWellRow(new NonNegativeInteger(0), 0, 0);
+          store.setWellColumn(new NonNegativeInteger(0), 0, 0);
+      }
+   
+      store.setWellExternalIdentifier(cellh5WellCoord, 0, 0);
+      
+      String site_id = MetadataTools.createLSID("WellSample", 0);
+      store.setWellSampleID(site_id, 0, 0, 0);
+      store.setWellSampleIndex(NonNegativeInteger.valueOf(seriesSite.get(s)), 0, 0, 0);
+      store.setWellSampleImageRef(getCurrentFile(), 0, 0, 0);
```
Comment (melissalinkert) [inline components/formats-gpl/src/loci/formats/in/CellH5Reader.java]: ```getCurrentFile()``` is not the correct value to pass in here; it needs to be a valid Image ID, e.g. ```MetadataTools.createLSID("Image", s)```.

File: components/formats-gpl/src/loci/formats/in/CellH5Reader.java [hunk 5]
```diff
@@ -481,8 +495,44 @@ private void parseStructure() throws FormatException {
...
+          //
+      }
+     
+      if (wellRowLetterIndex > -1 && wellColNumberIndex > 0) {
+          store.setWellRow(new NonNegativeInteger(wellRowLetterIndex), 0, 0);
+          store.setWellColumn(new NonNegativeInteger(wellColNumberIndex - 1), 0, 0);
+      } else {
+          store.setWellRow(new NonNegativeInteger(0), 0, 0);
+          store.setWellColumn(new NonNegativeInteger(0), 0, 0);
+      }
```
Comment (melissalinkert) [inline components/formats-gpl/src/loci/formats/in/CellH5Reader.java]: The Row and Column population looks much better - 0-based indexes are correct.

Do note though that this still only populates one Well and WellSample.  The last 2 parameters to ```store.setWell``` calls are the plate index and the well index, respectively; the well index needs to be incremented for each unique well.  Similarly for ```store.setWellSample``` calls, the last 3 parameters are the plate index, well index, and site (field) index.

**Comment by melissalinkert** [comment]
This now has the commits from gh-1761 - would you like us to review all of the changes together?  If so, we can close gh-1761 and proceed with reviewing the writer changes here.  If not, could you please remove the writer commits from this branch?

File: components/formats-gpl/src/loci/formats/in/CellH5Reader.java (hunk 5 above)
Comment (sommerc) [inline components/formats-gpl/src/loci/formats/in/CellH5Reader.java]: Yes, I thought so. Right now, I just write one image. Is it okay to set these indices to zero? Or, how do can I save all open images from ImageJ? With this, I could iterate and increment these indices appropriately.

**Comment by sommerc** [comment]
I close the CellH5Writer PR and pushed all commits to here. Thx again, for your help and insightful comments. I really learn a lot! :)

**Comment by sommerc** [comment]
any news?

**Comment by joshmoore** [comment]
I've kicked off travis again, but currently it's failing due to:
```
[ERROR] Failed to execute goal on project formats-gpl: Could not resolve dependencies for project ome:formats-gpl:jar:5.1.2-SNAPSHOT: Could not find artifact cisd:jhdf5:jar:14.12.0 in central (http://repo.maven.apache.org/maven2) -> [Help 1]
```

**Comment by sommerc** [comment]
It looks like that the appropriate JHDF5 14.12.0 artifact is not in the Artifactory... is it possible to add it there?

**Comment by melissalinkert** [comment]
The JHDF5 14.12.0 artifact is now deployed, with the group ID ```ch.systems.cisd``` as before.  You will need to update ```components/formats-gpl/pom.xml``` to revert the change to ```groupId```, as that should remain the same across version numbers.

**Comment by melissalinkert** [comment]
It looks like you may also need to merge the current 'develop' branch into this one, as there are some conflicts.

**Comment by sommerc** [comment]
thanks Melissa!

**Comment by sommerc** [comment]
Build looks better now: Only the java 8 maven build gives "Failed to collect dependencies at woolz:JWlz:jar:1.4.0". Have no idea, what's the problem there... :)

**Comment by sbesson** [comment]
@sommerc: thanks for the heads-up. The Jwlz dependency issue was due to the OME artifactory being temporarily unavailable. I restarted the Travis build which should fix the status.

**Comment by sommerc** [comment]
Thank you!!

**Comment by sommerc** [comment]
Any news?

**Comment by sommerc** [comment]
If there's still something missing / wrong?

**Comment by melissalinkert** [comment]
Including in the build; fine with me to merge if the next round of merge builds remain green.

I'm still a little concerned that the plate metadata isn't being populated as accurately as it could be, but we can address that in a follow-up PR if needed.

**Comment by melissalinkert** [comment]
ROIHandler changes should probably get a sign-off from one or more of @bramalingam/@jburel/@pwalczysko too though.

**Comment by jburel** [comment]
The ROIhandler changes will lead to rois on the wrong planes when exporting the images as ome-tiff/ome-xml
The ```loci.plugins.out.Exporter``` will have to take that change into account.
We can take care of it in a follow up PR.

**Comment by pwalczysko** [comment]
Yes, I can confirm, a ROI drawn in Fiji on Z 12 is now displayed on Z 13 in OMERO -> there is a shift by one which has to be taken care of.

**Comment by sommerc** [comment]
okay, I don't know what Z12/13 is. Is there sth for me to fix!?

**Comment by jburel** [comment]
@sommerc: I will open a PR to address that issue.

**Comment by jburel** [comment]
PR opened see https://github.com/sommerc/bioformats/pull/1

**Comment by pwalczysko** [comment]
In Fiji, 4 cases tested : 
  - 1. roi drawn in OMERO and opened in Fiji on specific z plane
  - 2. roi drawn in Fiji and saved to OMERO on specific z plane
  - 3. roi drawn in OMERO and opened in Fiji on specific t plane
  - 4. roi drawn in Fiji and saved to OMERO on specific t plane

Case 1. has a shift of +1 on Fiji side . 
In OMERO: 
screen shot 2015-08-05 at 11 06 01
In Fiji
screen shot 2015-08-05 at 11 05 41

Case 2. works as expected 
Case 3. has a shift of +1 on Fiji side . 
In OMERO
screen shot 2015-08-05 at 11 04 19
In Fiji
screen shot 2015-08-05 at 11 04 00
Case 4 does not work at all - all ROIs drawn in Fiji via ROI Manager on a specific t plane will import to OMERO as being on t = 1.  Double-checked with 5.1.3 release bioformats_package.jar in ImageJ and this is working there as expected => regression.


In summary, out of the 4 workflow cases 3 failed the test (regressions).

**Comment by jburel** [comment]
@pwalczysko:
sorry I should have clarified, the test to do is export as ome-tiff.

**Comment by sommerc** [comment]
sorry I am getting lost a bit here..., is this actually independent of the cellh5 reader&writer? Yet, I am not sure how to read-in the rois (from cellh5) so that they arrive correctly in fiji... In my test cellh5 reading created the correct roi assignments in Fiji

**Comment by jburel** [comment]
@sommerc: it is.
@pwalczysko is doing another test. The easiest strategy will be to merge as it is, and fix the roihandler in a follow up PR if any issue
I will discuss with @melissalinkert this afternoon

**Comment by pwalczysko** [comment]
@sommerc : No worries, my bad - I am just retesting this, as @jburel mentioned.

**Comment by pwalczysko** [comment]
Retest: After clarification with @jburel , testing just the Bio-formats exporter, not the OMERO side of things which will have to be sorted separately and has nothing to do with the changes here.

In ImageJ, 

1. exporting an image with ROI drawn on z plane 10, after export and reimport to ImageJ I have in Metadata:
```
<Shape ID="Shape:0:0" StrokeColor="-65281" Text="0019-0195-0164" TheC="0" TheT="0" TheZ="9">
```
and the ROI is shown as expected on z plane 10 (checked via ROI manager).

2. exporting an image with ROI drawn on t 29 and reimporting it I have
```
<Shape ID="Shape:0:0" Text="0057-0106-0106" TheC="0" TheT="28" TheZ="0">
```
and the image has the ROI on t plane 29 as expected when checked via ROI manager.

This looks like the expeced behaviour.

**Comment by jburel** [comment]
@pwalczysko: thanks. I should have been clearer to avoid confusion

**Comment by sommerc** [comment]
@jburel, @pwalczysko :) no worries! Thanks for your time and efforts on this!

**Comment by melissalinkert** [comment]
Merging as discussed with @jburel.  We'll open separate PRs to address the above issues, so not something you need to worry about @sommerc.

**Comment by sommerc** [comment]
Awesome! :-)
